controller/controlplane/payload_customizer.go

```diff
@@ -0,0 +1,46 @@
+package controlplane
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/pkg/telemetry/types"
+	"github.com/kong/kubernetes-telemetry/pkg/provider"
+)
+
+// defaultPayloadCustomizer creates a PayloadCustomizer that injects the hostname into the payload
+// and removes unnecessary fields.
+// TODO: Consider setting the hostname via an environment variable using the node name and the Kubernetes Downward API for improved configurability.
+func defaultPayloadCustomizer(hostnameRetriever func() (string, error)) (types.PayloadCustomizer, error) {
+	var hostname string
+	var err error
+	if hostnameRetriever != nil {
+		if hostname, err = hostnameRetriever(); err != nil {
+			return nil, fmt.Errorf("failed to get defaultPayloadCustomizer: %w", err)
+		}
+	} else {
+		// Get the hostname where KO is running.
+		if hostname, err = os.Hostname(); err != nil {
+			return nil, fmt.Errorf("failed to get defaultPayloadCustomizer: %w", err)
+		}
+	}
+
+	payloadCustomizer := func(payload types.Payload) types.Payload {
+		// Create a copy of the payload
+		newPayload := make(types.Payload, len(payload))
+		for k, v := range payload {
+			newPayload[k] = v
+		}
```

Comment:
I'm OK with us making a copy here (even if that's not strictly needed, I'm not sure).

In any case, we can use https://pkg.go.dev/maps#Clone to achieve this.

Reply:
The customizer changes the original payload, no need to copy the original map.